### DISCUSSION TITLE
Conscious Language in Avocado

### DISCUSSION THREAD

**Discussion by beraldoleal** [avocado-framework/avocado#4384] [RFC]
Hi all,

I hope you are doing fine during this time.

I'm seeing some projects [1] eradicating problematic language [2] (words like 'master', 'slave', 'black-list', 'white-list', etc) and I will be glad to start this effort of removing those words here but first I would like to know the Avocado's community option on this topic.

[1] - https://github.com/github/renaming
[2] - https://www.redhat.com/en/blog/making-open-source-more-inclusive-eradicating-problematic-language

**Comment by apahim** [comment]
+1, but no need to discuss that much really, IMO. I mean, considering that the matter makes the community less inclusive, people that are excluded by that aren't here to give their opinion, right? Any merit discussions won't have the full picture in that case.
Technically, it should be easy enough. I would recommend starting with a PR and keep the discussion in the PR review around the best new naming for each case.
Just my 2 cents.

**Comment by clebergnu** [comment]
I agreed there shouldn't be much to discuss indeed, and I don't think a patch to improve the wording situation, if it exists, would have any opposition.

I actually went through the Avocado source code a number of months ago, and found that `git grep $problematic_word` reveals a few things we could change.  c4f6a3a8f47d18e00f21e4f186723f697ca4148b is one example... but there is at least one that we can not change (`avocado/utils/multipath.py`) as it would break the interaction with the tool itself.

PS: one thing to keep in mind is that a single word may have multiple meanings, and not all are problematic.  For instance, the word "master" as a noun can indeed mean someone that owns or employs slaves, but as an adjective, attached to a noun it means something else entirely.  Examples include "master mason" meaning "of high skill" and "master branch" simply meaning "primary" or "most important". The Oxford Dictionary uses similar examples.

**Reply by beraldoleal** [reply]
@clebergnu  are you saying that we should keep the master branch? Because IMO associating "most import" to "master" it is a bad example.

https://sfconservancy.org/news/2020/jun/23/gitbranchname/

**Reply by clebergnu** [reply]
I'm saying that I don't think the word "master" (adjective) is interchangeable with the word (noun) associated with "slave", so this and other projects (and Git itself) IMO never had the intention to invoke that meaning.

Still, for practical purposes, be it for adhering to the new Git (and associated tools and services) default, or following sponsors' suggestions and guidelines, it should be changed.

**Reply by beraldoleal** [reply]
Thanks for the comments, @clebergnu.

I will try to keep it practical and I will not enter on the merit if they had the intention or not, IMO this is not the main subject of the discussion. Like @apahim said "people that are excluded by that aren't here to give their opinion, right?"

**Comment by pevogam** [comment]
Hi all,

Well, I disagree with a lot of the premises about the so called conscious language with an argumentation I already provided in the email threads (I am getting confused which avenue should we use for sharing thoughts now that we switch among emails, discussions, and sometimes still issues) but I guess the votes are against me here. I am providing the same argumentation here just for visibility and for consistency of these discussions on multiple locations:

"From a purely technical perspective I must say I am against this change. It is not in the
intent of the original creators of these terms to refer to any phenomena relating to human
society and other people or external party's choice to interpret the words out of their
original context should not be an obstacle in the workflow of the rest of us. I would like
to not lose from the role-descriptive power of these terms in order to adjust for such
misinterpretations.

I want to clarify that my views on this are purely technical and I am not attaching any
meaning to these other than the one they were conceived with in our workflows."

In any case I don't expect my arguments will change anybody's mind. I have had lots of academic and some industrial collaborations with international teams and I find the consideration of switching terms already harmful to many as it starts pointing unnecessary fingers instead of remaining formal and technically neutral but it is not my place to talk for everybody's emotional state when it comes to this.

**Reply by beraldoleal** [reply]
@pevogam your argument is based on the premise that the authors had no intentions about this relationship. 
But you are saying that your argument is "technical", so I'm confused here.

Git is "inspired" by CSV and BitKeeper. And the first appearance of the word "master" inside git is on a CSV script[1]). BitKeeper also uses the word "master" as the main branch[2], and assuming that this "master"  is not related to "master/slave" is a wrong assumption IMO, since BitKeeper has strong references to "master/slave" [3]. 

[1] - https://github.com/git/git/commit/3e91311ae750af9bf2e3517b1e701288ac3066b9
[2] - http://www.bitkeeper.org/tips.html#_how_do_i_rebase_my_work_on_top_of_a_different_changeset
[3] - https://github.com/bitkeeper-scm/bitkeeper/blob/master/doc/HOWTO.ask#L223

**Reply by pevogam** [reply]
> But you are saying that your argument is "technical", so I'm confused here.

These terms have a technical meaning about a role in a relationship model where one of two entities has managing role with respect to the other. The meaning is technical in the sense that it is expressive of functional roles and implementation. In your motivation it is taken out of context and directed to social and unrelated such.

> @pevogam your argument is based on the premise that the authors had no intentions about this relationship.

This is my main argument yes (and probably of many) but also the fact that the authors' terms should remain unrelated to such improper application. Modifying them is already adapting them to out-of-context influence which shouldn't be there on the first place.

> and assuming that this "master" is not related to "master/slave" is a wrong assumption IMO, since BitKeeper has strong references to "master/slave" [3].

I see and so we have words like client and "server" and plenty of others which could be misused and mishandled. Just like scientific terms that have a strict definition used more loosely or even negatively by other social groups, it doesn't mean that use in an unrelated field should influence and even override use within this field. People can pollute a terminology you use in your workflow constantly and I am not sure if you will be switching your workflow each time they decide to do it, especially if beyond your control.

**Reply by beraldoleal** [reply]
> > But you are saying that your argument is "technical", so I'm confused here.
> 
> These terms have a technical meaning about a role in a relationship model where one of two entities has managing role with respect to the other. The meaning is technical in the sense that it is expressive of functional roles and implementation. In your motivation it is taken out of context and directed to social and unrelated such.

There are plenty of threads right now on the Internet discussing the real origin of "master" in git, some are saying they are related to "master/slave" "relationship managing role" aged from BitKeeper (as I linked before), some are saying they are related to "master recording". I'm talking about core devs discussing the real meaning of the word, like Petr Baudis: 

_"I picked the names "master" (and "origin") in the early Git tooling back in 2005. (this probably means you shouldn't give much weight to my name preferences :) ) I have wished many times I would have named them "main" (and "upstream") instead. Glad it's happening natfriedman"_

They are changing the name, because of technical reasons, because of a bad choice.

And I agree with the ideas here, regardless of the intent and the original mean, words matter at the end, to the readers.

This is IMO a good technical reason for changing by itself: Today's default name is confusing and uses an analogy/metaphor instead of a specific word.

IMO, "main" is a better choice technically speaking. Git and GH are providing tooling to change the default branch name because of that.

Now, my motivation goes beyond the technical clarification, is not about the origin of the term or the intent neither. It is about today's usage and how this might be harmful to others. For me, the origin of the term is irrelevant in this discussion.
 
> > @pevogam your argument is based on the premise that the authors had no intentions about this relationship.
> 
> This is my main argument yes (and probably of many) but also the fact that the authors' terms should remain unrelated to such improper application. Modifying them is already adapting them to out-of-context influence which shouldn't be there on the first place.

Even when authors are trying to change that word and recognizing that was a bad choice?

Authors can make mistakes, they can be non-English speakers, they can be fresh and choose without thinking too much about the consequences. 

Language is constantly changing. If we get to the origin of some words, some are used today in a completely different mean of the original intent (take the word "awful" for instance). But like I said, authors are trying to change the default branch name, authors are saying this was a bad choice. 

> > and assuming that this "master" is not related to "master/slave" is a wrong assumption IMO, since BitKeeper has strong references to "master/slave" [3].

Here, I would like to fix myself: yes, there are strong references to master/slave on BitKeeper, but also some strong references to "master copy" as I linked above.

> I see and so we have words like client and "server" and plenty of others which could be misused and mishandled. Just like scientific terms that have a strict definition used more loosely or even negatively by other social groups, it doesn't mean that use in an unrelated field should influence and even override use within this field. People can pollute a terminology you use in your workflow constantly and I am not sure if you will be switching your workflow each time they decide to do it, especially if beyond your control.

Well, to me this is the definition of "kindness". You listen, and you believe in people. It is not our job to say "this is not so bad to me". But I think this is not the place to start a discussion about what compassion is.

We are talking about a branch name and some words. What is the real impact on your day-to-day activities that a "sed" and/or an "opsss the default branch has changed, now it is foo" cannot solve?

That was my original intent when started this thread: What are the "technical challenges" here? Where are we going to cause trouble with this change? Unless I'm missing something, IMO it is just to simple change, let do it and move on.

**Reply by pevogam** [reply]
> IMO, "main" is a better choice technically speaking. Git and GH are providing tooling to change the default branch name because of that.

This brings something new to the table indeed. +1 about this argument

> Now, my motivation goes beyond the technical clarification, is not about the origin of the term or the intent neither. It is about today's usage and how this might be harmful to others. For me, the origin of the term is irrelevant in this discussion.

Same for me, I am not referring to historical importance or anything like that - customs can be changed and people have to adapt to new standards as long as as the new standard is proven to be better in some sense.

> > This is my main argument yes (and probably of many) but also the fact that the authors' terms should remain unrelated to such improper application. Modifying them is already adapting them to out-of-context influence which shouldn't be there on the first place.
> 
> Even when authors are trying to change that word and recognizing that was a bad choice?

Yes, this historical argument is a bit duplicated with the new argument above but it makes sense.

> > I see and so we have words like client and "server" and plenty of others which could be misused and mishandled. Just like scientific terms that have a strict definition used more loosely or even negatively by other social groups, it doesn't mean that use in an unrelated field should influence and even override use within this field. People can pollute a terminology you use in your workflow constantly and I am not sure if you will be switching your workflow each time they decide to do it, especially if beyond your control.
> 
> Well, to me this is the definition of "kindness". You listen, and you believe in people. It is not our job to say "this is not so bad to me". But I think this is not the place to start a discussion about what compassion is.

If we are going to kidnap the topic let me say that I had retreated to a Buddhist monastery in the past and lived there for a while so I had a few contacts with compassion :) What I am referring to here though remains technical and I would prefer not to mix political or emotional views with unrelated engineering fields like this one. This being said, I understand that here we also deal with humans from the very beginning so whether we want it or not this has to be taken into account.

> We are talking about a branch name and some words. What is the real impact on your day-to-day activities that a "sed" and/or an "opsss the default branch has changed, now it is foo" cannot solve?
> 
> That was my original intent when started this thread: What are the "technical challenges" here? Where are we going to cause trouble with this change? Unless I'm missing something, IMO it is just to simple change, let do it and move on.

Regarding the technical challenges, what I could foresee is lots of denied pushes because people keep using "master" and other default naming for most of the repositories they contribute to, then perhaps some breakage in automation pipelines here and downstream (lots of us users have built automated environments around the current ecosystem), and a few more headaches but nothing that we cannot handle and restore from in the long term.

Is such a potential rename also planned for Avocado VT? I hope at least it could be unified across all related repos.

I also hope you still understood my previous point that *reacting to external context* of negative origin by adapting our own terminology is the same as acknowledging some legitimacy of this context. At least in a perfect world, such negative master/slave connotation should not even be given a chance to exist. But perhaps what I am saying is not easy to understand, the technical deficiencies as well as the arbitrariness of the author wording is what actually convinces me more here.

**Reply by beraldoleal** [reply]
@pevogam sorry for the delay here. I got preempted with other tasks and I just realized that we have an open question here:

> Is such a potential rename also planned for Avocado VT? I hope at least it could be unified across all related repos.

I agree with you here. And I would love to see this change on avocado-vt too. Although I think this initiative must come from a more active member in that repository. In any case, I will try to bring that topic there too when the time comes.

**Comment by willianrampazzo** [comment]
I found this link interesting for this discussion: https://inclusivenaming.org/faqs/

**Comment by marian-bucur** [comment]
Hi, I'm new here. I read https://www.redhat.com/en/blog/making-open-source-more-inclusive-eradicating-problematic-language. Could you change name if it unwelcome me? I plan to use avocado, but people call me avocado skin because of a disease and it makes me sad.
